common/src/java/org/apache/hadoop/hive/common/FileUtils.java

```diff
@@ -1387,6 +1388,47 @@ public static RemoteIterator<LocatedFileStatus> listFiles(FileSystem fs, Path pa
         status -> filter.accept(status.getPath()));
   }
 
+  /**
+   * Resolves a symlink on a local filesystem. In case of any exceptions or scheme other than "file"
+   * it simply returns the original path. Refer to DEBUG level logs for further details.
+   * @param path
+   * @param conf
+   * @return
+   * @throws IOException
+   */
+  public static Path resolveSymlinks(Path path, Configuration conf) throws IOException {
+    if (path == null) {
+      return null;
+    }
+
+    FileSystem srcFs;
+    String scheme = path.toUri().getScheme();
+    if (scheme != null) {
+      srcFs = path.getFileSystem(conf);
```

Comment:
if scheme isn't null & if it is not file, do we need to get the FS, we can return without getting the FS, right?

Reply:
good catch, we can return early in that case